geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+            .iter()
+            .cloned()
+            .chain(rings.into_iter().map(|p| p.exterior().clone()))
+            .collect::<Vec<_>>();
+        poly = Polygon::new(exterior, interiors);
+    }
+    poly
+}
+
+/// Inputs to this function is a unordered set of lines that must form a valid multipolygon
+fn stitch_multipolygon_from_lines<F: GeoFloat>(
+    lines: Vec<Line<F>>,
+) -> PolygonStitchingResult<MultiPolygon<F>> {
+    let rings = stitch_rings_from_lines(lines)?;
+
+    fn find_parent_idxs<F: GeoFloat>(
+        ring_idx: usize,
+        ring: &LineString<F>,
+        all_rings: &[LineString<F>],
+    ) -> Vec<usize> {
+        all_rings
+            .iter()
+            .enumerate()
+            .filter(|(other_idx, _)| ring_idx != *other_idx)
+            .filter_map(|(idx, maybe_parent)| {
+                Polygon::new(maybe_parent.clone(), vec![])
+                    .contains(ring)
+                    .then_some(idx)
+            })
+            .collect()
+    }
+
+    // Associates every ring with its parents (the rings that contain it)
+    let parents: HashMap<usize, Vec<usize>> = rings
+        .iter()
+        .enumerate()
+        .map(|(ring_idx, ring)| {
+            let parent_idxs = find_parent_idxs(ring_idx, ring, &rings);
+            (ring_idx, parent_idxs)
+        })
+        .collect();
+
+    // Associates outer rings with their inner rings
+    let mut polygons_idxs: HashMap<usize, Vec<usize>> = HashMap::default();
+
+    // For each ring, we check how many parents it has  otherwise it's an outer ring
+    //
+    // This is important in the scenarios of "donuts" where you have an outer donut shaped
+    // polygon which completely contains a smaller polygon inside its hole
+    for (ring_index, parent_idxs) in parents.iter() {
+        let parent_count = parent_idxs.len();
+
+        // if it has an even number of parents, it's an outer ring so we can just add it if it's
+        // missing
+        if parent_count % 2 == 0 {
+            polygons_idxs.entry(*ring_index).or_default();
+            continue;
+        }
+
+        // the direct parent is the parent which has itself the most parents
+        fn find_direct_parents(
+            parents_indexs: &[usize],
+            parents: &HashMap<usize, Vec<usize>>,
+        ) -> Option<usize> {
+            parents_indexs
+                .iter()
+                .filter_map(|parent_idx| {
+                    parents
+                        .get(parent_idx)
+                        .map(|grandparents| (parent_idx, grandparents))
+                })
+                .max_by_key(|(_, grandparents)| grandparents.len())
+                .map(|(idx, _)| idx)
+                .copied()
+        }
+
+        // if it has an odd number of parents, it's an inner ring
+        //
+        // to find the specific outer ring it is related to, we search for the direct parent.
+        if let Some(direct_parent) = find_direct_parents(parent_idxs, &parents) {
```

Comment:
I think this is just an artifact from the `find_direct_parents` function since it uses `max_by_key` which returns an `Option<T>`. It shouldn't be `None` in theory. I always prefer this since I'm shying away from panicing but that might be valid here. I don't know what to do here tbh

Reply:
I'm a big fan of debug_assert for stuff that "should never happen but I don't want to crash a production app over it".

Reply:
I think I was just a bit lazy yesterday. We can also just put a new error variant in place there.